flows/template.go

```diff
@@ -0,0 +1,113 @@
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/nyaruka/goflow/assets"
+	"github.com/nyaruka/goflow/utils"
+)
+
+// Template represents messaging templates used by channels types such as WhatsApp
+type Template struct {
+	assets.Template
+}
+
+// NewTemplate returns a new template objects based on the passed in asset
+func NewTemplate(t assets.Template) *Template {
+	return &Template{Template: t}
+}
+
+// Asset returns the underlying asset
+func (t *Template) Asset() assets.Template { return t.Template }
+
+// FindTranslation finds the matching translation for the passed in channel and languages (in priority order)
+func (t *Template) FindTranslation(channel assets.ChannelUUID, langs []utils.Language) *TemplateTranslation {
+	// first iterate through and find all translations that are for this channel
+	matches := make(map[utils.Language]assets.TemplateTranslation)
+	for _, tr := range t.Template.Translations() {
+		if tr.Channel().UUID == channel {
+			matches[tr.Language()] = tr
+		}
+	}
+
+	// now find the first that matches our language
+	for _, lang := range langs {
+		tr := matches[lang]
+		if tr != nil {
+			return NewTemplateTranslation(tr)
+		}
+	}
+
+	return nil
+}
+
+// TemplateTranslation represents a single translation for a template
+type TemplateTranslation struct {
+	assets.TemplateTranslation
+}
+
+// NewTemplateTranslation returns a new TemplateTranslation for the passed in asset
+func NewTemplateTranslation(t assets.TemplateTranslation) *TemplateTranslation {
+	return &TemplateTranslation{TemplateTranslation: t}
+}
+
+// Asset returns the underlying asset
+func (t *TemplateTranslation) Asset() assets.TemplateTranslation { return t.TemplateTranslation }
+
+var templateRegex = regexp.MustCompile(`({{\d+}})`)
+
+// Substitute substitutes the passed in variables in our template
+func (t *TemplateTranslation) Substitute(vars []string) string {
+	s := string(t.Content())
+	for i, v := range vars {
+		s = strings.ReplaceAll(s, fmt.Sprintf("{{%d}}", i+1), v)
+	}
+
+	// replace any remaining unmatched items
+	s = templateRegex.ReplaceAllString(s, "")
+
+	return s
+}
+
+// TemplateAssets is our type for all the templates in an environment
+type TemplateAssets struct {
+	templates []*Template
+	byName    map[string]*Template
+}
+
+// NewTemplateAssets creates a new template list
+func NewTemplateAssets(ts []assets.Template) *TemplateAssets {
+	templates := make([]*Template, len(ts))
+	byName := make(map[string]*Template)
+	for i, t := range ts {
+		template := NewTemplate(t)
+		templates[i] = template
+		byName[t.Name()] = template
+	}
+
+	return &TemplateAssets{
+		templates: templates,
+		byName:    byName,
+	}
+}
+
+// FindTranslation looks through our list of templates to find the template matching the passed in name
+// If no template or translation is found then empty string is returned
+func (l *TemplateAssets) FindTranslation(name string, channel *assets.ChannelReference, langs []utils.Language) *TemplateTranslation {
```

Comment:
wouldn't it be better to lean on UUID here instead of name? I'm just trying think where we might go in the future with templates, and it seems possible in the future that they are editable in RP and thus the UUID is the immutable identifier of a template, and name is the thing that might be stale in a flow definition.

Reply:
Man, I feel we are just making work for ourselves with these uuids but sure I will change this.

Reply:
Maybe.. or maybe we're future proofing ourselves to using templates differently